Alerting: Fix flaky SQLITE_BUSY when migrating with provisioned dashboards

JacobsonMT opened the pull request:
This PR moves the new migration service from background service `Run(...)` to server `Init`.

**Background**

SQLite database write contention between the migration's single transaction and dashboard provisioning's frequent commits was causing the migration to consistently with `SQLITE_BUSY`/`SQLITE_BUSY_SNAPSHOT` on all retries.

This is not a new issue for SQLite+Grafana, but the discrepancy between the length of the transactions was causing it to be very consistent. In addition, since a failed migration has implications on the assumed correctness of the alertmanager and alert rule definition state, we cause a server shutdown on error. This can make e2e tests as well as some high-load provisioned SQLite installations flaky on startup.

**Fix**

The fix implemented is more avoiding the issue than fixing the root problem. We move the migration from a background service `Run(...)` to server `Init`. Since dashboard provisioning also happens in a background service `Run(...)`, depending on timing and length of transactions it was intermittently but frequently failing migration commits by invalidating the read transaction snapshot (for WAL). Moving the migration alongside the other provisioning calls means it will always run before dashboard provisioning and thus will not be contending with its writes.

The correct fix for this is better transaction management across various services but is out of scope for this change as we're primarily interested in mitigating the current bout of server failures in e2e tests when using SQLite.

**Steps to reproduce**

To reproduce locally (before this fix):

- Define some dashboards as provisioned files.
- Use SQLite as a db backend with the following configs:
  - `[database]`
    - `type=sqlite3`
    - `wal=true`
    - leave `max_open_conn` undefined or set it to a really high number.
  - `[unified_alerting]`
    - enabled = true
- Now start grafana server with a fresh database (delete or move `grafana.db`, `grafana.db-shm`, `grafana.db-wal`)
- It should fail fairly frequently.

JacobsonMT committed:
Alerting: Move migration from background service to server init

sqlite database write contention between the migration's single transaction and
dashboard provisioning's frequent commits was causing the migration to
 fail with SQLITE_BUSY/SQLITE_BUSY_SNAPSHOT on all retries.

 This is not a new issue for sqlite+grafana, but the discrepancy between the
 length of  the transactions was causing it to be very consistent. In addition,
 since a failed migration has implications on the assumed correctness of the
 alertmanager and alert rule definition state, we cause a server shutdown on
 error. This can make e2e tests as well as some high-load provisioned
 sqlite installations flaky on startup.

 The correct fix for this is better transaction management across various
 services and is out of scope for this change as we're primarily interested in
 mitigating the current bout of server failures in e2e tests when using sqlite.

papagian reviewed (commented):
What if we run the migration during the ngalert service [startup](https://github.com/grafana/grafana/blob/faa22b8f2067fe752af7a16c5605306bd2ef81a6/pkg/services/ngalert/ngalert.go#L47) (as we do for data migrations in other services) or the ngalert DB store [startup](https://github.com/grafana/grafana/blob/faa22b8f2067fe752af7a16c5605306bd2ef81a6/pkg/services/ngalert/store/database.go#L46) instead of having it called directly by the server?

eleijonmarck commented on pkg/services/ngalert/migration/service.go:
```diff
@@ -51,7 +51,7 @@ func ProvideService(
 // If the migration status in the kvstore is set and both unified alerting is disabled and ForceMigration is set to true, the migration will be reverted.
 func (ms *MigrationService) Run(ctx context.Context) error {
 	var errMigration error
-	errLock := ms.lock.LockExecuteAndRelease(ctx, actionName, time.Minute*10, func(context.Context) {
```
ah this seems to be it. the ctx was not passed along

papagian replied:
I think it does pass [here](https://github.com/grafana/grafana/blob/e4c1a7a141683604e035fa1a07fde975fde30d8b/pkg/infra/serverlock/serverlock.go#L63)

grobinson-grafana replied:
Do the migrations still need to be moved from `Run` to `Init` then?

JacobsonMT replied:
This is not relevant to the fix, just small improvements I came across while figuring this out, clean as you go and all that. I can remove them and add them to the next migration-related PR, if you prefer.

JacobsonMT commented:
> What if we run the migration during the ngalert service [startup](https://github.com/grafana/grafana/blob/faa22b8f2067fe752af7a16c5605306bd2ef81a6/pkg/services/ngalert/ngalert.go#L47) (as we do for data migrations in other services) or the ngalert DB store [startup](https://github.com/grafana/grafana/blob/faa22b8f2067fe752af7a16c5605306bd2ef81a6/pkg/services/ngalert/store/database.go#L46) instead of having it called directly by the server?

I would heavily prefer this solution and it was originally how I tried structuring it. The issue I came across was that you don't have access to guardian from wire provide methods, so I can't use any service methods that need guardian (for example, `folderService.Get`). Basically, I hit a `panic: no guardian factory implementation provided`. If there's a way around this, I'd go that route.

papagian commented:
> > What if we run the migration during the ngalert service [startup](https://github.com/grafana/grafana/blob/faa22b8f2067fe752af7a16c5605306bd2ef81a6/pkg/services/ngalert/ngalert.go#L47) (as we do for data migrations in other services) or the ngalert DB store [startup](https://github.com/grafana/grafana/blob/faa22b8f2067fe752af7a16c5605306bd2ef81a6/pkg/services/ngalert/store/database.go#L46) instead of having it called directly by the server?
> 
> I would heavily prefer this solution and it was originally how I tried structuring it. The issue I came across was that you don't have access to guardian from wire provide methods, so I can't use any service methods that need guardian (for example, `folderService.Get`). Basically, I hit a `panic: no guardian factory implementation provided`. If there's a way around this, I'd go that route.

AFAIK guardian is set [here](https://github.com/grafana/grafana/blob/0ddc3d4a49790ba0cb50ed34fbfb8fdfc8226631/pkg/services/guardian/provider.go#L21)
so I think that it you would be resolved if you added guardian as an ngalert dependency
but @grafana/grafana-authnz-team may have more context

JacobsonMT committed:
Move migration run to ngalert wire init instead of server

requires some structure changes to allow for a fake so that unrelated alerting
tests don't run the migration.

JacobsonMT commented:
Great suggestion @papagian, the guardian injection worked so I moved it to ngalert wire init instead of server.

papagian reviewed (approved):
I have tested and alert rules are migrated as expected
